tests/service.rs

```diff
@@ -66,3 +66,42 @@ fn restart_after_ungraceful_shutdown_should_succeed() {
         client.verify_pod_condition(&pod, "Ready");
     }
 }
+
+// This test provokes race conditions but does not guarantee their
+// absence on success.
+#[test]
+fn no_race_conditions_should_occur_if_many_pods_are_started_and_stopped_in_parallel() {
+    let mut client = TestKubeClient::new();
+    client.timeouts().verify_pod_condition = Duration::from_secs(120);
+
+    setup_repository(&client);
+
+    let pod_spec = indoc! {"
+        apiVersion: v1
+        kind: Pod
+        metadata:
+          name: agent-service-integration-test-race-condition
+        spec:
+          containers:
+            - name: noop-service
+              image: noop-service:1.0.0
+              command:
+                - noop-service-1.0.0/start.sh
+          tolerations:
+            - key: kubernetes.io/arch
+              operator: Equal
+              value: stackable-linux
+    "};
+
+    let mut pods = Vec::new();
+
+    for _ in 1..=100 {
+        pods.push(TemporaryResource::new(&client, &with_unique_name(pod_spec)));
+    }
+
+    for pod in &pods {
+        client.verify_pod_condition(&pod, "Ready");
+    }
+
+    pods.clear();
```

Comment:
I may be missing the point of this test, but doesn't the interesting part start after clearing the Vec<_> when all pods go out of scope and get deleted?
 
I would have expected a check after this to see if all pods get deleted successfully and none remain in state `terminating` due to the agent missing a stopped signal.*/

Something along these lines: 

```rust
let podnames = pods
        .iter()
        .map(|pod| pod.metadata.name.unwrap())
        .collect::<Vec<_>>();

pods.clear();

for podname in podnames {
    assert!(client.verify_pod_is_deleted(&podname));
}
```

Reply:
After thinking about it some more, that would probably just be the same test again, as missing a signal would lead to the pod never becoming ready?

Reply:
The functions in `test::kube::KubeClient` listen for the according events.

For instance, `delete` awaits the `Deleted` event within the specified duration.

```
/// Timeouts for operations
pub struct Timeouts {
    pub delete: Duration,
}
```

```
    /// Deletes the given resource and awaits the confirmation of the deletion.
    pub async fn delete<K>(&self, resource: K) -> Result<()>
```

`TestKubeClient` panics if an operation was not successful and the test fails:

```
/// A client for interacting with the Kubernetes API
///
/// [`TestKubeClient`] is a synchronous version of [`KubeClient`] which
/// additionally panics on erroneous results. It reduces the verbosity of
/// test cases.
pub struct TestKubeClient {
    
    /// Deletes the given resource.
    pub fn delete<K>(&self, resource: K)
```

So, `delete` means "delete the resource or panic" and not "try to delete the resource and let the caller verify that the operation did what it was supposed to do".

Reply:
I rewrote the test case so that this conversation does not apply anymore and could be closed.